src/tree/hash_input.rs

```diff
@@ -55,44 +109,136 @@ impl<'a> ParentNodeHashInput<'a> {
             right_hash,
         }
     }
-    pub fn hash(&self, ciphersuite: &Ciphersuite) -> Vec<u8> {
+    pub(crate) fn hash(&self, ciphersuite: &Ciphersuite) -> Vec<u8> {
         let payload = self.encode_detached().unwrap();
         ciphersuite.hash(&payload)
     }
 }
 
-pub struct LeafNodeHashInput<'a> {
-    node_index: &'a NodeIndex,
-    key_package: &'a Option<KeyPackage>,
-}
+// === Parent hashes ===
 
-impl<'a> LeafNodeHashInput<'a> {
-    pub fn new(node_index: &'a NodeIndex, key_package: &'a Option<KeyPackage>) -> Self {
-        Self {
-            node_index,
-            key_package,
+/// The list of HPKEPublicKey values of the nodes in the resolution of `index`
+/// but with the `unmerged_leaves` of the parent node omitted.
+pub(crate) fn original_child_resolution(
+    tree: &RatchetTree,
+    index: NodeIndex,
+) -> Vec<&HPKEPublicKey> {
+    // Build the exclusion list that consists of the unmerged leaves of the parent
+    // node
+    let mut unmerged_leaves = vec![];
+    // If the current index is not the root, we collectthe unmerged leaves of the
+    // parent
+    if let Ok(parent_index) = treemath::parent(index, tree.leaf_count()) {
+        // Check if the parent node is not blank
+        if let Some(parent_node) = &tree.nodes[parent_index].node {
+            for index in &parent_node.unmerged_leaves {
+                unmerged_leaves.push(NodeIndex::from(*index as usize));
+            }
         }
-    }
-    pub fn hash(&self, ciphersuite: &Ciphersuite) -> Vec<u8> {
-        let payload = self.encode_detached().unwrap();
-        ciphersuite.hash(&payload)
-    }
+    };
+    // Convert the exclusion list to a HashSet for faster searching
+    let exclusion_list: HashSet<&NodeIndex> = HashSet::from_iter(unmerged_leaves.iter());
+
+    // Compute the resolution for the index with the exclusion list
+    let resolution = tree.resolve(index, &exclusion_list);
+
+    // Build the list of HPKE public keys by iterating over the resolution
+    resolution
+        .iter()
+        .map(|index| tree.nodes[*index].public_hpke_key().unwrap())
+        .collect()
 }
 
-impl<'a> Codec for ParentNodeHashInput<'a> {
-    fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
-        self.node_index.encode(buffer)?;
-        self.parent_node.encode(buffer)?;
-        encode_vec(VecSize::VecU8, buffer, &self.left_hash)?;
-        encode_vec(VecSize::VecU8, buffer, &self.right_hash)?;
-        Ok(())
+/// Computes the parent hashes for a leaf node and returns the parent hash for
+/// the parent hash extension
+pub(crate) fn compute_parent_hashes(tree: &mut RatchetTree, index: LeafIndex) -> Vec<u8> {
```

Comment:
Is there a particular reason that this is not part of `impl RatchetTree`? Same goes for `compute_tree_hash`.